proxy/src/deshred.rs

```diff
@@ -0,0 +1,194 @@
+
+use std::sync::atomic::AtomicBool;
+use crossbeam_channel::Receiver;
+
+pub struct WrappedShred(Shred);
+
+impl Ord for WrappedShred {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.index().cmp(&other.0.index())
+    }
+}
+
+impl PartialOrd for WrappedShred {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.index().partial_cmp(&other.0.index())
+    }
+}
+
+impl PartialEq for WrappedShred {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.index() == other.0.index()
+    }
+}
+
+impl Eq for WrappedShred {}
+
+
+pub fn insert_packet_into_shred_bucket(packet: &Packet, shred_bucket_by_slot:  &Arc<RwLock<HashMap<Slot, BTreeSet<WrappedShred>>>>, entry_sender: &Arc<Sender<DeshredEntry>>) {
+    let shred = packet_to_shred(packet);
+    if shred.is_none() {
+        return;
+    }
+    let shred = shred.unwrap(); 
+    let slot = shred.slot();
+
+    if !shred.is_data(){
+        return;
+    }
+
+    // debug!("shred slot {:?} idx {:?}", slot, shred.index());
+    // create a new slot bucket if it doesn't exist, else insert the shred into the existing slot bucket
+    let mut shred_bucket_by_slot = shred_bucket_by_slot.write().unwrap();
+    let slot_bucket  = shred_bucket_by_slot.get(&slot);
+    if slot_bucket.is_none() {
+        shred_bucket_by_slot.insert(slot, BTreeSet::new());
+    }
+    let slot_bucket = shred_bucket_by_slot.get_mut(&slot).unwrap();
+    slot_bucket.insert(WrappedShred(shred));
+
+
+    let last_shred = slot_bucket.last().unwrap(); 
+    if last_shred.0.last_in_slot()  && slot_bucket.len() == (last_shred.0.index() + 1) as usize{ 
+        debug!("deshredding slot {:?}", slot);
+        let shreds = slot_bucket.iter().map(|shred| shred.0.clone()).collect_vec(); 
+        
+        let data = Shredder::deshred(shreds.as_slice());
+        if data.is_err() {
+            debug!("failed to deshred shreds {:?}", data.err());
+            return;
+        }
+        let data = data.unwrap();
+        let entries: Result<Vec<Entry>, _> = bincode::deserialize(&data);
+        if entries.is_ok() {
+            let entries = entries.unwrap();
+            debug!("deshredded entries {:?}", entries);
+            if entry_sender.receiver_count() > 0 {
+                for entry in entries {
+                    let send_result = entry_sender.send(DeshredEntry{
+                        num_hashes: entry.num_hashes,
+                        hash: entry.hash.to_bytes().to_vec(),
+                        // tbd: is bincode supported in other languages ?
+                        transactions: entry.transactions.iter().map(|t|  bincode::serialize(t).unwrap()).collect_vec(), 
+                    });
+                    if send_result.is_err() {
+                        error!("failed to send deshredded entry to deshred server {:?}", send_result.err());
+                    }
+                }
+            }
+            shred_bucket_by_slot.remove(&slot);
```

Comment:
Potential memory leak here if a slot bucket never fills. Not sure if that ever happens and how common it is. 

Simple solution would be to drop the bucket after every interval regardless of it been filled.